rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+# RESTful Service Specifications
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Goals
+
+ - Uniformity between data-centric APIs and remote procedure APIs
+ - Use of pre-existing standards/conventions when possible
+ - Minimization of bike-shedding
+
+Standards/Conventions used as inspiration:
+ - [Whitehouse API Standards](https://github.com/WhiteHouse/api-standards)
+  - RESTful URLs
+ - [Spring Data REST](http://docs.spring.io/spring-data/rest/docs/current/reference/html/)
+  - paging conventions
+  - uses [JSON HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08)
+ - [JSend](http://labs.omniti.com/labs/jsend)
+  - API response conventions
+
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
```

Comment:
> I'm tempted to say let's simplify by disallowing projection. 

My instinct is to be against this. We have seen performance gains in consumers of lightblue by being able to finely tune responses.

> There's a possible middle-ground solution where we can provide pre-set projections rather than allowing whoever to project however they want.

What about allowing multiple pre-set projections to be layered?